// omnichat-model/built/Schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var OmniChatSchema;
(function (OmniChatSchema) {
    let MessageType;
    (function (MessageType) {
        MessageType["Image"] = "IMAGE";
        MessageType["Text"] = "TEXT";
        MessageType["Audio"] = "AUDIO";
        MessageType["File"] = "FILE";
        MessageType["Video"] = "VIDEO";
        MessageType["Location"] = "LOCATION";
        MessageType["Product"] = "PRODUCT";
        MessageType["Variant"] = "VARIANT";
        MessageType["PaymentRequest"] = "PAYMENT_REQUEST";
        MessageType["OrderConfirmation"] = "ORDER_CONFIRMATION";
        MessageType["OrderStatus"] = "ORDER_STATUS";
        MessageType["Collection"] = "COLLECTION";
    })(MessageType = OmniChatSchema.MessageType || (OmniChatSchema.MessageType = {}));
    let MessageStatus;
    (function (MessageStatus) {
        MessageStatus["OutgoingCreated"] = "OUTGOING_CREATED";
        MessageStatus["OutgoingSent"] = "OUTGOING_SENT";
        MessageStatus["OUTGOING_DELIVERED"] = "OUTGOING_DELIVERED";
        MessageStatus["OutgoingRead"] = "OUTGOING_READ";
        MessageStatus["OutgoingError"] = "OUTGOING_ERROR";
        MessageStatus["OutgoingSilent"] = "OUTGOING_SILENT";
        MessageStatus["IncomingRecevied"] = "INCOMING_RECEIVED";
        MessageStatus["IncomingRead"] = "INCOMING_READ"; // Read by the salesPerson
    })(MessageStatus = OmniChatSchema.MessageStatus || (OmniChatSchema.MessageStatus = {}));
    let ChatPlatform;
    (function (ChatPlatform) {
        ChatPlatform["Whatsapp"] = "WHATSAPP";
        ChatPlatform["FBMessenger"] = "FBMESSENGER";
        ChatPlatform["Skype"] = "SKYPE";
        ChatPlatform["Email"] = "EMAIL";
        ChatPlatform["Sms"] = "SMS";
        ChatPlatform["Chatclub"] = "CHATCLUB";
    })(ChatPlatform = OmniChatSchema.ChatPlatform || (OmniChatSchema.ChatPlatform = {}));
    let CustomerSupportRequestType;
    (function (CustomerSupportRequestType) {
        CustomerSupportRequestType["Button"] = "BUTTON";
        CustomerSupportRequestType["Collection"] = "COLLECTION";
        CustomerSupportRequestType["Chat"] = "CHAT";
    })(CustomerSupportRequestType = OmniChatSchema.CustomerSupportRequestType || (OmniChatSchema.CustomerSupportRequestType = {}));
    let CustomerSupportRequestStatus;
    (function (CustomerSupportRequestStatus) {
        CustomerSupportRequestStatus["Waiting"] = "WAITING";
        CustomerSupportRequestStatus["Contacted"] = "CONTACTED";
        CustomerSupportRequestStatus["Finished"] = "FINISHED";
    })(CustomerSupportRequestStatus = OmniChatSchema.CustomerSupportRequestStatus || (OmniChatSchema.CustomerSupportRequestStatus = {}));
    // export type CustomerGender = "male" | "female" | "neutral" | "other"
    let CustomerGender;
    (function (CustomerGender) {
        CustomerGender["Male"] = "male";
        CustomerGender["Female"] = "female";
    })(CustomerGender = OmniChatSchema.CustomerGender || (OmniChatSchema.CustomerGender = {}));
})(OmniChatSchema = exports.OmniChatSchema || (exports.OmniChatSchema = {}));
