// omnichat-model/built/User.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Parse = require("parse/node");
var UserStatus;
(function (UserStatus) {
    UserStatus["LoogedIn"] = "LOGGEDIN";
    UserStatus["LoogedOut"] = "LOGGEDOUT";
    UserStatus["Away"] = "AWAY";
})(UserStatus = exports.UserStatus || (exports.UserStatus = {}));
var WebPushEvent;
(function (WebPushEvent) {
    WebPushEvent["Chat"] = "CHAT";
    // Only the orders that I have created
    WebPushEvent["OrderStatus"] = "ORDER_STATUS";
    // All retailer orders, must be admin
    WebPushEvent["OrderStatusAll"] = "ORDER_STATUS_ALL";
})(WebPushEvent = exports.WebPushEvent || (exports.WebPushEvent = {}));
class User extends Parse.User {
    constructor(attributes) {
        super('User', attributes);
    }
    get email() { return this.getEmail(); }
    set email(value) { this.setEmail(value); }
    get name() { return this.get('name'); }
    set name(value) { this.set('name', value); }
    get lastName() { return this.get('lastName'); }
    set lastName(value) { this.set('lastName', value); }
    get retailer() { return this.get('retailer'); }
    set retailer(value) { this.set('retailer', value); }
    /* This user is only for API access */
    get apiUser() { return this.get('apiUser'); }
    set apiUser(value) { this.set('apiUser', value); }
    get whatsAppNumber() { return this.get('whatsAppNumber'); }
    set whatsAppNumber(value) { this.set('whatsAppNumber', value); }
    get status() { return this.get('status'); }
    set status(value) { this.set('status', value); }
    get pricingTable() { return this.get('pricingTable'); }
    set pricingTable(value) { this.set('pricingTable', value); }
    get webPushEndpoint() { return this.get('webPushEndpoint'); }
    set webPushEndpoint(value) { this.set('webPushEndpoint', value); }
    get webPushEvents() { return this.get('webPushEvents'); }
    set webPushEvents(value) { this.set('webPushEvents', value); }
}
exports.User = User;
Parse.Object.registerSubclass('User', User);
